refactor: ♻️ include dependency management in values, not just in principles (#261)
seedcase-project/design: software/index.qmd

title: "Software"
---

These design documents describe our overall software design, including how we develop software and how are software products are designed.

So far, these documents include our values towards software design, described
below, and our [principles](principles.qmd), which are informed by our values.

## Values

{{< include /includes/_values-preamble.qmd >}}

Our values, and the things we believe in, direct and shape how we build software. We believe that these values facilitate work within the team and with external collaborations.

- Developing software with inclusivity and accessibility in mind produces
  better, more usable software.

- Software that is open source and uses open source tools is better in the long
  term and more likely to be used.

- Documentation is as important as the software itself.

- High-quality code, though partially subjective, is as important and valuable
  as software that works.

- Transparent and clear software is easier to reason about, build, and use.

- Simplicity requires more effort but makes better software.

- Building software that incorporates sustainability, in the broad sense for both

>   humans and the environment, produces better software.

Author: ```suggestion
- Sustainability in software design is not only good for humans and the environment, 
  but also for the software itself.
```

- Fewer, carefully selected third-party dependencies produces better
  software that is more stable, reliable, and with fewer security and

>   maintenance issues.

Author: ```suggestion
- Incorporate and consider security early and review often, 
  such as for dependencies, as it is more important than ever.
```

- Prioritising both the design of software and the quality of the code itself

>   will mean fewer breaking changes.

Author: ```suggestion
- Reliability in stable software, such as minimizing breaking changes, is about respecting users.
```

So far, these documents include our values towards software design, described
below, and our [principles](principles.qmd), which are informed by our values.

## Values

{{< include /includes/_values-preamble.qmd >}}

Our values, and the things we believe in, direct and shape how we build
software. We believe that these values facilitate work within the team and with
external collaborations.

- Developing software with inclusivity and accessibility in mind produces
  better, more usable software.

- Software that is open source and uses open source tools is better in the long
  term and more likely to be used.

- Documentation is as important as the software itself.

- High-quality code, though partially subjective, is as important and valuable
  as software that works.

- Transparent and clear software is easier to reason about, build, and use.

- Simplicity requires more effort but makes better software.

- Sustainability in software design is not only good for humans and the
  environment, but also for the software itself.

- Fewer, carefully selected third-party dependencies produces better software
  that is more stable, reliable, and with fewer security and

> 

Author: ```suggestion
```

Oops forgot to remove

title: "Software"
---

These design documents describe our overall software design, specifically on how
the software will generally look like in the end and our general approach to how

> we build and develop it.

Reviewer: ```suggestion
These design documents describe our overall software design, including how we develop software and how are software products are designed.
```

So far, these documents include our values towards software design, described
below, and our [principles](principles.qmd), which are informed by our values.

## Values

{{< include /includes/_values-preamble.qmd >}}

Our values, or things we believe in, direct and shape how we build our software

> and are:

Reviewer: ```suggestion
Our values, and the things we believe in, direct and shape how we build software. We believe that these values facilitate work within the team and with external collaborations.
```

Author: Small edit, but I want to keep the "believe" part.

> - Developing software with inclusivity and accessibility in mind produces
>   better, more usable software.

Reviewer: ```suggestion
- Developing software with inclusivity and accessibility in mind produces
  more usable software for everyone.
```

Author: Value are things we *believe* in and they don't require empirical proof. That's why I used "better" because I believe it will be better. (it's hard to prove inclusivity makes things better, but I believe/want to live in a world where that is something that is considered, so for me, it is better).

Reviewer: Hmm, I didn't mentioned empirical proof anywhere. My point is about explaining how/why we believe something is better, e.g. "we believe focusing on accessibility makes software more usable for everyone" rather than just saying "accessibility is better".

I think these explanations actually makes our value page itself more accessible since it is now easier to understand for a new reader who does not already share the same notion of "better" as the author of the page.

Author: Could you read the principles page first before getting too technical/verbose here? Principles are where we can be more explicit/descriptive. Values aren't meant to be super precise. For example, not everyone has the same definition of "more usable". More usable *to who*? If it was just about more usable, we'd be building plugins for Excel, since more people would value that as more usable. But that's not what we value or want. For us, that's not "more usable", nor "better".

If you want to read more about value statements, etc, check out e.g. https://asana.com/resources/company-values-examples or other examples. Values are short, simple, concise. They can even be a few words or even just one. E.g. Meta's value for a long time was "move fast and break things".

Reviewer: I've read the principle page and I think it's great that we are elaborating more in detail there; I'm not trying to replace the principle page. My suggestion is to avoid vague statements when we can be more explicit since that makes our values clearer and easier to understand.

I don't think I'm suggesting something radical here. Many of the other values are written in the format I'm suggesting, e.g. further down we have "Transparent and clear software is easier to reason about, build, and use". We didn't just write "Transparent and clear software is better".

Likewise, most of the values from asana in the article you linked are written as a single word, and then a short explanation why they believe in that value, which is in line with what I suggest here. An example:

> Mindfulness. We focus on the present and aim to give ourselves time to reflect and space to integrate what we learn. These practices allow us to collectively learn from and improve all that we do, and to continually evolve our culture. 

Overall I don't quite understand the resistance. To me, the suggestions is a minor change that makes the statement easier to understand for a new reader by clarifying what we mean (unless you meant something very different from what I wrote with "better"; then please add that). The cost of the change is one extra word which neither seems too technical nor too verbose. Could you explain what you think is lost with the change I suggest here?

Author: The resistance is mainly for two reasons: 1) this PR isn't about those other values and 2) I don't want more verbose values, I want less. Your suggestions added a lot more words. If you can give suggestions, in another PR, with changes that reduce words, not add them, I'd probably have less resistance. Would you like to make a PR with some changes?

Author: And to expand on other examples, see [Zen of Python](https://peps.python.org/pep-0020/) and [Agile Manifesto](https://agilemanifesto.org). These are my inspirations for our values.

Reviewer: The reason I included the comments here is because there were changes to all the value statements in the PR diff. Normally I would say yes to creating a new PR, but I think I'm too confused about the requirements to make one in this case:

- Regarding verbosity: For the suggestion I made in this thread, there is one more word added. On the other hand for the new value added in this PR, I suggested removing a few words to make it more succinct, but they were added back in. I also suggested to remove some seemingly duplicated values to make things more succinct.
- Regarding vagueness: Currently I think the values on this page are a mix of vague and explicitly explained and I don't understand when to use one or the other. For example, the new one added in this PR contains an explanation:
    > Fewer, carefully selected third-party dependencies produces better software that is more stable, reliable, and with fewer security and maintenance issues.
   
    That's an elaboration of why we prefer fewer dependencies, which I think made it clearer than if we would simple write "Fewer dependencies is better".

So overall, it seems like I'm missing something and what seems inconsistent and vague to me is in in line with how you want the values to be written and interpreted. That's ok, I just don't think I could make a PR to clarify anything and we might waste time going in circles, so I'll approve this as is. I could try to understand this more on a call in the future, but it seems like it might take some time, and there are other higher priorities items to focus on first.

Author: Yea, to be fair, I also don't love what I wrote and will come back to them at some point in the future to simplify them. But they were worming away in my mind and needed them out, and couldn't think of a more concise/succinct way of writing them than that. And I mean, maybe "fewer dependencies is better" is actually a better value 🤔 And yea, it might help to have a call about it at some point 👍

Author: But you've got a worm in my head now to simplify/streamline these more 😝 in another PR that is.

> - Software that is open source and uses open source tools is better in the long
>   term and more likely to be used.

Reviewer: ```suggestion
- Software that is open source and uses open source tools is easier to build, maintain, adopt, and understand.
```

Author: Same comment as the above. The principles page is where more precise language is used. Values can be vague, and in fact, values are vague as they are high-level abstractions of human things.

Reviewer: Same reply as above. I think we actually have specific things in mind when we say "better in the long term" and being explicit about them benefits both us and other readers of this page.

- Documentation is as important as the software itself.

- High-quality code, though partially subjective, is as important and valuable
  as software that works.

> - Transparent and clear software is easier to reason about, build, and use.
> 

Reviewer: ```suggestion
```

Duplicate with the open source point above

Author: Open source doesn't mean transparent, nor does it mean clear.

Reviewer: If this is not related to the software being easier to understand as in the open source point above, then I think this needs further elaboration here to show what is distinct.

Author: See the principles for more explanation. Values are not meant to be long expositions. For example, Stormlight Archive fantasy series is about saying "ideals" and live by them, which leads to more powerful magic https://stormlightarchive.fandom.com/wiki/Immortal_Words These words are vague and can mean many things. That's the purpose of them. One of the first "words" is: "Life before death. Strength before weakness. Journey before destination." Very vague. But so simple and conveys so many things. "Life before death" sounds like obvious.. but think about it more, it's a lot more complex and complicated. 

Or, another example, in the movie "Hero" with Jet Li. The emperor is given a symbol for the word "Sword" (劍) by an assassin. The assassin is wanting to see if they emperor can interpret the other meanings of Sword. It's a really beautiful film and about how even simple words convey a lot more meaning than their first glance might seem.

All this to say, our values are meant to be like that. Vaguer, with many layers of meaning attached to them, rather than surgical precision.

Also, I guess you haven't read the organisation values section either, it is like this page :stuck_out_tongue:

Reviewer: I honestly don't think that "These words are vague and can mean many things" sounds useful in an organization... It's great with catchy slogans, but there if there are no explanations it becomes a bit like a secret club of only us knowing what we mean with our values (or in this case, not even I understand).

That doesn't mean I'm against a catchy slogan or value statement; they have inspiring qualities like you pointing out. I just don't think it takes anything away from a value statement to explain what is meant by it.

Author: Hmm, they aren't meant to be catchy slogans. They are values. I don't know how to be more clear than that. Values don't need to be explained to be meaningful and useful, especially to help guide all the other stuff. Values are like primitive compasses. They help give a general direction, but they can't give a detailed roadmap. Values are a bit like vision statements.

Reviewer: Yeah, I think it's always useful to explain the *why*, even for values since there are reasons we picked these values and not others. Sounds like an interesting future phone call! Actually I fondly remember us talking about this briefly a few years ago on those bridges in Copenhagen =)

> - Simplicity requires more effort but makes better software.

Reviewer: ```suggestion
- Simplicity requires more effort but makes software that is more sustainable.
```

Author: same as above.

Reviewer: Same reply as above here too, just for visibility so we don't forget about this one.

- Software that is built with a consideration of sustainability in the broad
  sense lasts longer and is easier to build and maintain.

> 

Reviewer: ```suggestion
```

Isn't this point a platitude? We're saying that software that is built to be sustainable ends up being more sustainable.

Author: Hmm, I see what you mean. I'll revise it slightly:

```suggestion
- Building software that incorporates sustainability, in the broad sense for both
  humans and the environment, produces better software.

```

- Fewer, but carefully vetted third-party dependencies produces products with
  fewer security and maintenance issues, and better reliability and stability

>   overall.

Reviewer: ```suggestion
- Fewer, carefully selected third-party dependencies produces better 
  software that is more stable, reliable, and with fewer security and 
  maintenance issues.
```

Author: Small edits to your edits ☺️

- Prioritising both the design of software and the quality of the code itself

>   will mean fewer breaking changes, which we want to avoid.

Reviewer: ```suggestion
- Prioritising both the design of software and the quality of the code itself
  will mean fewer breaking changes.
```

The way it was written it sounded like we want to avoid fewer breaking changes =p

- Building software with these values makes it easier to build as a team and

>   with collaborators.

Reviewer: ```suggestion
```

Since this is not a value in itself, I moved it to the short intro paragraph.
